src/rawchannel.cc

```diff
@@ -312,41 +330,103 @@ class RawChannel : public Nan::ObjectWrap
+        }
       }
-    }
-
-    int flags = 0;
 
-    if (hw->m_NonBlockingSend)
-      flags = MSG_DONTWAIT;
-
-    int i = send(hw->m_SocketFd, &frame, sizeof(struct can_frame), flags);
+      int flags = 0;
 
+      if (hw->m_NonBlockingSend)
+        flags = MSG_DONTWAIT;
+    int i = send(hw->m_SocketFd, &frame, sizeof(struct can_frame), flags);  
     info.GetReturnValue().Set(i);
   }
 
+ /**
+   * Send a CAN message immediately.
+   *
+   * PLEASE NOTE: By default, this function may block if the Tx buffer is not available. Please use
+   * createRawChannelWithOptions({non_block_send: false}) to get non-blocking sending activated.
+   *
+   * Added by Guillaume Tournabien 2021_03_14
+   * Note : All the setup is not supported and no protection are included
+   * 
+   * @method send
+   * @param message {Object} JSON object describing the CAN message, keys are id, length, data {Buffer}, ext or rtr
+   */
+  static NAN_METHOD(SendFD)
+  {
+    int requiered_mtu = 1;
+    RawChannel* hw = ObjectWrap::Unwrap<RawChannel>(info.Holder());
+
+    CHECK_CONDITION(info.Length() >= 1, "Invalid arguments");
+    CHECK_CONDITION(info[0]->IsObject(), "First argument must be an Object");
+    CHECK_CONDITION(hw->IsValid(), "Invalid channel!");
+    struct canfd_frame frameFD;                         // GT modif
+    frameFD.flags = 0;                                  //GT modif : Flags not used
+
+    v8::Local<v8::Context> context = info.GetIsolate()->GetCurrentContext();
+    v8::Local<v8::Object> obj = Nan::To<Object>(info[0]).ToLocalChecked();
+
+    // CAN FD FRAME
+    // ---------------
+    // TODO: Check for correct structure of message object
+
+      frameFD.can_id = obj->Get(context, id_symbol).ToLocalChecked()->ToUint32(context).ToLocalChecked()->Value();
+      if (obj->Get(context, ext_symbol).ToLocalChecked()->IsTrue())
+        frameFD.can_id  |= CAN_EFF_FLAG;
+
+      v8::Local<v8::Value> dataArg = obj->Get(context, data_symbol).ToLocalChecked();
+
+      CHECK_CONDITION(node::Buffer::HasInstance(dataArg), "Data field must be a Buffer");
+
+      // Get frame data
+      frameFD.len = node::Buffer::Length(Nan::To<Object>(dataArg).ToLocalChecked()); 
+      memcpy(frameFD.data, node::Buffer::Data(Nan::To<Object>(dataArg).ToLocalChecked()), frameFD.len);
+      
+      { // set time stamp when sending data
+        struct timeval now;
+        if ( 0==gettimeofday(&now, 0)) {
+          Nan::Set(obj, tssec_symbol, Nan::New((int32_t)now.tv_sec));
+          Nan::Set(obj, tsusec_symbol, Nan::New((int32_t)now.tv_usec));
+        }
+      }
+
+      int flags = 0;
+
+      if (hw->m_NonBlockingSend)
+      flags = MSG_DONTWAIT;
+      
+      /* ensure discrete CAN FD length values 0..8, 12, 16, 20, 24, 32, 64 */
+      // TODO : Need to implemenat stuffing function 
+      
+      // END TODO
+      int i = send(hw->m_SocketFd,&frameFD,sizeof(struct canfd_frame),requiered_mtu);   //GT modif works for CAN_FD frame   
```

Comment:
why are you passing 1 as a flag? Whats the purpose of required_mtu? It is set once but never change.

Reply:
Sorry but no real reason. The can-utils library use this flag on the cansend.c to send CAN_FD frame.
I reuse the code from can-utils but it not used on this function because on the cansend.c it's an argument that we can define.
I don't treat the MTU size but it's define on the structure that the max value of the CANFD_MTU is 72. (refercence can.h with linux 3.6)
I don"t think that it's critical but I will add information into TODO list